// @nightlylabs/connect/lib/nightlyConnect/connect/sdk/src/types/ClientMessages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConnectMessage = exports.ClientMessageTypes = void 0;
const AppMessages_1 = require("./AppMessages");
var ClientMessageTypes;
(function (ClientMessageTypes) {
    ClientMessageTypes["Connect"] = "Connect";
    ClientMessageTypes["GetInfo"] = "GetInfo";
    ClientMessageTypes["GetPendingRequests"] = "GetPendingRequests";
    ClientMessageTypes["ResolveSignTransaction"] = "ResolveSignTransaction";
    ClientMessageTypes["ResolveSignMessage"] = "ResolveSignMessage";
    ClientMessageTypes["RejectRequest"] = "RejectRequest";
})(ClientMessageTypes = exports.ClientMessageTypes || (exports.ClientMessageTypes = {}));
const createConnectMessage = (publicKey, sessionId, token, metadata) => {
    return {
        type: ClientMessageTypes.Connect,
        version: AppMessages_1.Version.V1,
        publicKey: publicKey,
        sessionId: sessionId,
        token: token,
        metadata
    };
};
exports.createConnectMessage = createConnectMessage;
