// @nightlylabs/connect/lib/nightlyConnect/connect/client-lib/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NCSolanaWalletAdapter = exports.ClientSolana = exports.AppSolana = exports.ClientNear = exports.AppNear = exports.NightlyConnectModal = void 0;
var NightlyConnectModal_1 = require("./modal/NightlyConnectModal");
Object.defineProperty(exports, "NightlyConnectModal", { enumerable: true, get: function () { return NightlyConnectModal_1.NightlyConnectModal; } });
var appNear_1 = require("../../sdk/src/near/appNear");
Object.defineProperty(exports, "AppNear", { enumerable: true, get: function () { return appNear_1.AppNear; } });
var clientNear_1 = require("../../sdk/src/near/clientNear");
Object.defineProperty(exports, "ClientNear", { enumerable: true, get: function () { return clientNear_1.ClientNear; } });
var appSolana_1 = require("../../sdk/src/solana/appSolana");
Object.defineProperty(exports, "AppSolana", { enumerable: true, get: function () { return appSolana_1.AppSolana; } });
var clientSolana_1 = require("../../sdk/src/solana/clientSolana");
Object.defineProperty(exports, "ClientSolana", { enumerable: true, get: function () { return clientSolana_1.ClientSolana; } });
__exportStar(require("../../sdk/src/types/AppMessages"), exports);
__exportStar(require("../../sdk/src/types/AppRequest"), exports);
__exportStar(require("../../sdk/src/types/ClientMessages"), exports);
__exportStar(require("../../sdk/src/types/AppMessages"), exports);
__exportStar(require("../../sdk/src/types/AppInfo"), exports);
__exportStar(require("../../../../common/types"), exports);
var NCSolanaWalletAdapter_1 = require("./adapter/NCSolanaWalletAdapter");
Object.defineProperty(exports, "NCSolanaWalletAdapter", { enumerable: true, get: function () { return NCSolanaWalletAdapter_1.NCSolanaWalletAdapter; } });
